File: Distinct Coloring.java
class Solution{
    public static long distinctColoring(int n, int[]r, int[]g, int[]b){
        //code here
        int arr[][] = new int[n][3];
        for(int i = 0;i<n;i++){
            arr[i][0] = r[i];
            arr[i][1] = g[i];
            arr[i][2] = b[i];
        }
        long dp[][] = new long[n][3];
        for(long a[] : dp) Arrays.fill(a, -1);
        long ans = Math.min(Math.min(util(0, 0,  arr, dp), util(0, 1,  arr, dp)), util(0, 2,  arr, dp));
        return ans;
    }
    public static long util(int i, int curr, int[][] arr, long dp[][]){
        if(i == arr.length) return 0;
        if(dp[i][curr] != -1) return dp[i][curr];
        long p = 0, q = 0;
        if(curr == 0){
            p= util(i+1, 1, arr, dp)+arr[i][1];
            q = util(i+1, 2, arr, dp)+arr[i][2];
        }
        if(curr == 1){
            p = util(i+1, 0, arr, dp)+arr[i][0];
            q = util(i+1, 2, arr, dp)+arr[i][2];
        }
        if(curr == 2){
            p = util(i+1, 0, arr, dp)+arr[i][0];
            q = util(i+1, 1, arr, dp)+arr[i][1];
        }
        return  dp[i][curr] = Math.min(p, q);
    }
}

File: Distinct Coloring.cpp
class Solution{   
public:
    long long int distinctColoring(int n, int r[], int g[], int b[]){
        // code here 
        vector<long long int>red(n,0);
        vector<long long int>green(n,0);
        vector<long long int>black(n,0);
        for(int i=0;i<n;i++){
            if(i==0){
                red[i]=r[i];
                green[i]=g[i];
                black[i]=b[i];
            }else{
                red[i]=min(r[i]+black[i-1],r[i]+green[i-1]);
                green[i]=min(g[i]+black[i-1],g[i]+red[i-1]);
                black[i]=min(b[i]+red[i-1],b[i]+green[i-1]);
            }
        }
        return min(green[n-1],min(red[n-1],black[n-1]));
    }
};
